numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -3229,15 +3316,40 @@ can_cast_fields_safety(PyArray_Descr *from, PyArray_Descr *to)
         }
         PyArray_Descr *to_base = (PyArray_Descr*)PyTuple_GET_ITEM(to_tup, 0);
 
-        NPY_CASTING field_casting = PyArray_GetCastSafety(from_base, to_base, NULL);
+        NPY_CASTING field_casting = PyArray_GetCastInfo(
+                from_base, to_base, NULL, &field_view_off);
         if (field_casting < 0) {
             return -1;
         }
         casting = PyArray_MinCastSafety(casting, field_casting);
+
+        /* Adjust the "view offset" by the field offsets: */
+        if (field_view_off != NPY_MIN_INTP) {
+            npy_intp to_off = PyLong_AsSsize_t(PyTuple_GET_ITEM(to_tup, 1));
+            if (error_converting(to_off)) {
+                return -1;
+            }
+            npy_intp from_off = PyLong_AsSsize_t(PyTuple_GET_ITEM(from_tup, 1));
+            if (error_converting(from_off)) {
+                return -1;
+            }
+            field_view_off = field_view_off - to_off + from_off;
+        }
+
+        /*
+         * If there is one field, use its field offset.  After that propagate
+         * the view offset if they match and set to "invalid" if not.
+         */
+        if (i == 0) {
```

Comment:
Could use a similar logic for the unstructured to structured case, to support the (rather ridiculous) case of multiple fields at the same offset...

Reply:
Yeah, I thought about it, and then thought that overlapping fields are ridiculous and should probably be discouraged (use array-views instead).  So had decided to not do it, but could look into it.